fw/http.c

```diff
@@ -1320,6 +1478,112 @@ tfw_http_send_resp(TfwHttpReq *req, int status, const char *reason)
 		tfw_h1_send_resp(req, status);
 }
 
+static void
+tfw_http_send_resp2(TfwHttpReq *req, TfwStr *msg, int status)
+{
+	if (TFW_MSG_H2(req)) {
+		tfw_h2_send_resp2(req, msg, status, 0);
+	} else {
+		TfwCliConn *cli_conn = (TfwCliConn *)req->conn;
+
+		WARN_ONCE(!list_empty_careful(&req->msg.seq_list),
+			  "Request is already in seq_queue\n");
+		tfw_stream_unlink_msg(req->stream);
+		spin_lock(&cli_conn->seq_qlock);
+		list_add_tail(&req->msg.seq_list, &cli_conn->seq_queue);
+		spin_unlock(&cli_conn->seq_qlock);
+
+		tfw_h1_send_resp2(req, msg, status);
+	}
+}
+
+static void
+tfw_http_req_redir(TfwHttpReq *req, int status, const char *pattern)
+{
+	size_t url_len = strlen(pattern) + 1;
+	char *url = kmalloc(url_len, GFP_KERNEL);
+	char *date_val = *this_cpu_ptr(&g_buf);
+	char *url_p = url;
+	const char *pattern_end = pattern + strlen(pattern);
+	TfwStr *c, *end;
+	char *status_line;
+
+	tfw_http_prep_date(date_val);
+
+#define TFW_STRCPY(from)                                                 \
+do {                                                                     \
+	size_t dif = url_p - url;                                            \
+	url = krealloc(url, url_len + (from)->len, GFP_KERNEL);              \
+	if (!url) {                                                          \
+		T_WARN("HTTP: unable to allocate memory for redirection url\n"); \
+		return;                                                          \
+	}                                                                    \
+	url_p = url + dif;                                                   \
+	TFW_STR_FOR_EACH_CHUNK(c, (from), end) {                             \
+		memcpy_fast(url_p, c->data, c->len);                             \
+		url_p += c->len;                                                 \
+	}                                                                    \
+} while (0)
+
+	for (; *pattern; ++pattern) {
+		if (*pattern == '$') {
+			size_t len = pattern_end - pattern - 1;
+
+			if (pattern_end - pattern > SLEN("request_uri") &&
+				!strncmp(pattern + 1, "request_uri", min(SLEN("request_uri"), len)))
+			{
+				TFW_STRCPY(&req->uri_path);
+				pattern += SLEN("request_uri");
+			} else if (pattern_end - pattern > SLEN("host") &&
+				       !strncmp(pattern + 1, "host", min(SLEN("host"), len)))
+			{
+				TFW_STRCPY(&req->host);
+				pattern += SLEN("host");
```

Comment:
I don't see any bug in the code, but I tested `uri == "/blog/" -> 301 = https://$host$request_uri;` - `$request_uri` does work, but `$host` is always empty. Maybe we need to consider which host to use - from URI or from Host header (you use the first one, the second one is stored in the headers table).